### PR TITLE
Support JSON as a serialization target

### PR DESCRIPTION
**Pull request by justinchuby**
### Description

Create serializers to the `json` format and register the `json` and `onnxjson` extensions to it.

```python
import onnx
model = onnx.load_model("onnx/backend/test/data/light/light_resnet50.onnx")
onnx.save_model(model, "resnet50.json")
```

yields

```json
{
  "ir_version": "3",
  "producer_name": "onnx-caffe2",
  "producer_version": "",
  "domain": "",
  "model_version": "0",
  "doc_string": "",
  "graph": {
    "node": [
      {
        "input": [
          "gpu_0/conv1_w_0__SHAPE"
        ],
        "output": [
          "gpu_0/conv1_w_0"
        ],
        "op_type": "ConstantOfShape",
        "attribute": [
          {
            "name": "value",
            "t": {
              "dims": [
                "1"
              ],
              "data_type": 1,
              "float_data": [
                0.02
              ],
              "name": ""
            },
            "type": "TENSOR"
          }
        ]
      },
      {
        "input": [
          "gpu_0/pred_b_0__SHAPE"
        ],
        "output": [
          "gpu_0/pred_b_0"
        ],
        "op_type": "ConstantOfShape",
        "attribute": [
          {
            "name": "value",
            "t": {
              "dims": [
                "1"
              ],
              "data_type": 1,
              "float_data": [
                0.02
              ],
              "name": ""
            },
            "type": "TENSOR"
          }
        ]
      },
      {
        "input": [
          "gpu_0/pred_w_0__SHAPE"
        ],
        "output": [
          "gpu_0/pred_w_0"
        ],
        "op_type": "ConstantOfShape",
        "attribute": [
          {
            "name": "value",
            "t": {
              "dims": [
                "1"
              ],
              "data_type": 1,
              "float_data": [
                0.02
              ],
              "name": ""
            },
            "type": "TENSOR"
          }
        ]
      },
...
```

### Motivation and Context

A more familiar text representation for free from protobuf using https://googleapis.dev/python/protobuf/latest/google/protobuf/json_format.html

### CODE REVIEW

**Review by jcwchen — COMMENTED** [review]
Thanks for the enhancement! I was wondering does JSON serialization/deserialization have 2GB limit as well?

**Comment by justinchuby** [comment]
I don't think there is one, but it's also not realistic to have big models in json because the text representation is less efficient.

**Comment by justinchuby** [comment]
onnxjson? onnx.json?

onnx.json is easier for editors
onnxjson is easier when we check extensions and work with paths.

I can add both json and onnxjson.

File: onnx/serialization.py [hunk 1]
```diff
@@ -141,7 +142,31 @@ def deserialize_proto(self, serialized: bytes | str, proto: _Proto) -> _Proto:
         return google.protobuf.text_format.Parse(serialized, proto)
 
 
+class _JsonSerializer(ProtoSerializer):
+    """Serialize and deserialize JSON."""
+
+    supported_format = "json"
+    file_extensions = frozenset({".json"})
```
Comment (justinchuby) [inline onnx/serialization.py:149]:
```suggestion
    file_extensions = frozenset({".json", ".onnxjson"})
```

**Review by liqunfu — APPROVED** [review]

**Comment by snnn** [comment]
> Thanks for the enhancement! I was wondering does JSON serialization/deserialization have 2GB limit as well?

Internally protobuf still use int instead of sizes_t as array length type.

**Pull request merged**